### PR TITLE
jclklib: Fix proxy calls popen() in hot path issue

### PR DESCRIPTION
**Pull request by JunAnnLaiIntel**
The issue is the proxy calls popen() to check whether the ptp4l is running or not, which causing high cpu usage.
This patch is to remove the popen() checking method and replace by checking there is packet available for the uds socket.

This patch is tested as below:
1. Proxy able to reconnect to ptp4l after ptp4l is dead and running again
2. The cpu usage is significantly low when setting the taskset for proxy program running on cpu 3. 
-LD_LIBRARY_PATH=$TEST_PATH $SCRIPT_PATH/jclklib_proxy
+LD_LIBRARY_PATH=$TEST_PATH taskset 0x8 $SCRIPT_PATH/jclklib_proxy
3. Set both summary_interval and logSyncInterval to -7 in igc.cfg file

image

image

### CODE REVIEW

File: jclklib/proxy/connect_ptp4l.cpp [hunk 1]
```diff
@@ -268,43 +258,42 @@ void *ptp4l_event_loop( void *arg)
...
-
-                    sk = m_sk.get();
-                    event_subscription(NULL);
-                    break;
+            } else {
+                /* Reset TIME_STATUS_NP data when ptp4l is disconnected */
+                pe.master_offset = 0;
+                memset(pe.gm_identity, 0, sizeof(pe.gm_identity));
+                pe.synced_to_primary_clock = false;
+                pe.as_capable = 0;
```
Comment (yoongsiang2) [inline jclklib/proxy/connect_ptp4l.cpp:273]: Add a printError here saying loss connection to ptp4l. Reset all data.

File: jclklib/proxy/connect_ptp4l.cpp [hunk 2]
```diff
@@ -268,43 +258,42 @@ void *ptp4l_event_loop( void *arg)
...
+
+                std::string uds_address;
+                m_sk.reset(sku);
+                uds_address = "/var/run/ptp4l";
+                if(!sku->setDefSelfAddress() || !sku->init() ||
+                        !sku->setPeerAddress(uds_address))
+                    fprintf(stderr, "Fail to connect to ptp4l\n");
+
+                sk = m_sk.get();
+                event_subscription(NULL);
```
Comment (yoongsiang2) [inline jclklib/proxy/connect_ptp4l.cpp:296]: Since msg_send() in line 277 is successful, do we still need line 283-295?

**Review by yoongsiang2 — CHANGES_REQUESTED** [review]

File: jclklib/proxy/connect_ptp4l.cpp [hunk 3]
```diff
@@ -268,43 +258,42 @@ void *ptp4l_event_loop( void *arg)
...
+                memset(pe.gm_identity, 0, sizeof(pe.gm_identity));
+                pe.synced_to_primary_clock = false;
+                pe.as_capable = 0;
+                notify_client();
+
+                while(1){
+                    if(msg_send(false))
+                        break;
+                    PrintError("Failed to connect to ptp4l. Retrying...");
+                    sleep(2);
```
Comment (AzuraTarmiziIntel) [inline jclklib/proxy/connect_ptp4l.cpp:280]: A question here, are we catching any Control-C for proxy ?

File: jclklib/proxy/connect_ptp4l.cpp [hunk 4]
```diff
@@ -331,15 +320,10 @@ int Connect::connect()
...
-    while (is_ptp4l_running() != 1) {
-        /* sleep for 2 seconds and keep looping until there is ptp4l available */
-        PrintError("Failed to connect to ptp4l. Retrying...");
-        sleep(2);
-    }
-    PrintInfo("Connected to ptp4l via /var/run/ptp4l.");
     pe.ptp4l_id = 1;
-
-    uds_address = "/var/run/ptp4l";
+    std::string interface = "enp1s0";
```
[the remark this answers is not in the document]
Comment (AzuraTarmiziIntel) [reply jclklib/proxy/connect_ptp4l.cpp:324]: any option to pass this interface down to proxy ?

File: jclklib/proxy/connect_ptp4l.cpp [hunk 5]
```diff
@@ -268,43 +258,31 @@ void *ptp4l_event_loop( void *arg)
...
+
+                /* Reset TIME_STATUS_NP data when ptp4l is disconnected */
+                pe.master_offset = 0;
+                memset(pe.gm_identity, 0, sizeof(pe.gm_identity));
+                pe.synced_to_primary_clock = false;
+                pe.as_capable = 0;
+                notify_client();
+
+                while(1){
+                    if(msg_send(false))
```
Comment (yoongsiang2) [inline jclklib/proxy/connect_ptp4l.cpp:279]: i think can straight away replace msg_send() with event_subscription().

**Review by yoongsiang2 — CHANGES_REQUESTED** [review]

File: jclklib/proxy/connect_ptp4l.cpp (hunk 3 above)
Comment (yoongsiang2) [reply jclklib/proxy/connect_ptp4l.cpp:282]: yes, this is in another parallel thread. termination signal is controlled in main thread. 
Maybe we can do some experiment here: kill ptp4l, let the proxy goes into subscript looping, then ctrl+c to see whether proxy can be terminated or not.

Comment (JunAnnLaiIntel) [reply jclklib/proxy/connect_ptp4l.cpp:282]: yes, when proxy is looping and retrying to connect to ptp4l, proxy is able to terminate with "ctrl + c".

**Comment by JunAnnLaiIntel** [comment]
Since this patch has been merged in [https://github.com/intel-staging/libptpmgmt_iaclocklib/pull/103](https://github.com/intel-staging/libptpmgmt_iaclocklib/pull/103 ), this PR will be closed.

**Comment by JunAnnLaiIntel** [comment]
Since this patch has been merged in https://github.com/intel-staging/libptpmgmt_iaclocklib/pull/103, this PR will be closed.

**Pull request closed without merging**